app/components/markdown/markdown_image.js

```diff
@@ -1,70 +1,231 @@
         this.state = {
-            width: 10000,
-            maxWidth: 10000,
-            height: 0
+            width: 0,
+            height: 0,
+            maxWidth: Math.MAX_INT,
+            failed: false
         };
+
+        this.mounted = false;
     }
 
     componentWillMount() {
-        Image.getSize(this.props.src, this.handleSizeReceived, this.handleSizeFailed);
+        this.loadImageSize(this.props.source);
+    }
+
+    componentDidMount() {
+        this.mounted = true;
     }
 
     componentWillReceiveProps(nextProps) {
-        if (this.props.src !== nextProps.src) {
-            Image.getSize(nextProps.src, this.handleSizeReceived, this.handleSizeFailed);
+        if (this.props.source !== nextProps.source) {
+            this.setState({
+                width: 0,
+                height: 0,
+                failed: false
+            });
+
+            this.loadImageSize(nextProps.source);
         }
     }
 
+    componentWillUnmount() {
+        this.mounted = false;
+    }
+
+    loadImageSize = (source) => {
+        Image.getSize(source, this.handleSizeReceived, this.handleSizeFailed);
+    };
+
     handleSizeReceived = (width, height) => {
+        if (!this.mounted) {
+            return;
+        }
+
         this.setState({
             width,
             height
         });
     };
 
     handleSizeFailed = () => {
+        if (!this.mounted) {
+            return;
+        }
+
         this.setState({
-            width: 0,
-            height: 0
+            failed: true
         });
-    }
+    };
 
     handleLayout = (event) => {
+        if (!this.mounted) {
+            return;
+        }
+
         this.setState({
             maxWidth: event.nativeEvent.layout.width
         });
-    }
+    };
+
+    handleLinkPress = () => {
+        const url = normalizeProtocol(this.props.linkDestination);
+
+        Linking.canOpenURL(url).then((supported) => {
+            if (supported) {
+                Linking.openURL(url);
+            }
+        });
+    };
+
+    handleLinkLongPress = async () => {
+        const {formatMessage} = this.context.intl;
+
+        const config = await mattermostManaged.getLocalConfig();
+
+        let action;
+        if (config.copyAndPasteProtection !== 'true') {
+            action = {
+                text: formatMessage({id: 'mobile.markdown.link.copy_url', defaultMessage: 'Copy URL'}),
+                onPress: this.handleLinkCopy
+            };
+        }
+
+        this.props.onLongPress(action);
+    };
+
+    handleLinkCopy = () => {
+        Clipboard.setString(this.props.linkDestination);
+    };
 
     render() {
-        let {width, maxWidth, height} = this.state; // eslint-disable-line prefer-const
+        let image = null;
+
+        if (this.state.width && this.state.height && this.state.maxWidth) {
+            let {width, height} = this.state;
+
+            if (Platform.OS === 'android' && (width > ANDROID_MAX_WIDTH || height > ANDROID_MAX_HEIGHT)) {
```

Comment:
I'm ok with this for now, but maybe we should look at resizing the image somehow